Add SquareCB contextual bandit algorithm

canondetortugas opened the pull request:
This pull request adds an implementation of the SquareCB contextual bandit algorithm (from this paper [https://arxiv.org/abs/2002.04926](https://arxiv.org/abs/2002.04926)) under cb_explore_adf.

jackgerrits commented:
Hi @canondetortugas, thanks for opening the PR! Can you please add some tests into RunTests? This is important to ensure this reduction continues to work as the codebase evolves.

canondetortugas commented:
Hey @jackgerrits, thanks for the quick response. I added some very basic unit tests to RunTests (essentially the same as what was there for RegCB) and verified that all the tests pass locally.

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+
+  float l = 0;
+  float u = maxw;
+  float w, v;
```
Nit: one declaration per line

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+  if (maxw * fhat * fhat <= delta)
+    return maxw;
+
+  float l = 0;
```
Can you add a comment explaining where these terms come from?

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+  uint32_t num_actions = (uint32_t)preds.size();
+
+  // The actual parameter $\gamma$ used in the SquareCB.
+  const float gamma = _gamma_scale*pow(_counter, _gamma_exponent); 
```
```suggestion
  const float gamma = _gamma_scale * std::pow(_counter, _gamma_exponent); 
```

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+  // RegCB action set parameters
+  const float max_range = _max_cb_cost - _min_cb_cost;
+  // threshold on empirical loss difference
+  const float delta = _c0 * log((float)(num_actions * _counter)) * pow(max_range, 2);
```
```suggestion
  const float delta = _c0 * std::log((float)(num_actions * _counter)) * std::pow(max_range, 2);
```

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+	  float min_cost = preds[0].score;
+	  // Compute highest-scoring action
+	  for(size_t a = 0; a < num_actions; ++a)
+	    {
```
Seems like the formatting is a bit off here, running `clang-format` on this file should fix it.

jackgerrits commented on vowpalwabbit/cb_explore_adf_squarecb.cc:
```diff
@@ -0,0 +1,348 @@
+  using config::make_option;
+  bool cb_explore_adf_option = false;
+  bool squarecb = false;
+  const std::string mtr = "mtr";
+  std::string type_string(mtr);
```
```suggestion
  std::string type_string = "mtr";
```

canondetortugas commented:
Cool, I addressed your comments and merged. It looks like it's failing the Windows CI and the formatting check though. Is there a short answer for how I should proceed with fixing these?

olgavrou commented:
Hi @canondetortugas including the new files `cb_explore_adf_squarecb.h` and `cb_explore_adf_squarecb.cc` in `vw_core.vcxproj` [here](https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/vw_core.vcxproj#L136) and [here](https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/vw_core.vcxproj#L267) respectively, should be enough for the windows ci to build.

Regarding the formatting, it is more of a nice-to-have than a strong requirement but if you want to resolve the formatting issues running `clang-format -i vowpalwabbit/cb_explore_adf_regcb.cc` and restoring the whitespace changes on the `cb_explore_adf_regcb.cc` file (there seems to be no other change on the latter file) should do the trick.

canondetortugas commented:
@olgavrou Thanks! Everything is passing now.

canondetortugas commented:
Added a wiki page here: https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Contextual-Bandit-Exploration-with-SquareCB
